fetcher/fetcher.go

```diff
@@ -0,0 +1,151 @@
+package fetcher
+
+import (
+	"archive/zip"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"github.com/dcos/dcos-diagnostics/dcos"
+	"github.com/sirupsen/logrus"
+)
+
+// EndpointFetchRequest is a struct passed to Fetcher with information about URL to be fetched
+type EndpointFetchRequest struct {
+	URL      string
+	Node     dcos.Node
+	FileName string
+}
+
+// FetchStatusUpdate is an update message published by Fetcher when EndpointFetchRequest is done. If error occurred	 during
+// fetch then Error field is not nil.
+type FetchStatusUpdate struct {
+	URL   string
+	Error error
+}
+
+// FetchBulkResponse is a message published when Fetcher finish it job due to cancelled context or closed input chanel
+type FetchBulkResponse struct {
+	ZipFilePath string
+}
+
+// Fetcher is a struct responsible for fetching nodes endpoints
+type Fetcher struct {
+	file         *os.File
+	client       *http.Client
+	input        <-chan EndpointFetchRequest
+	statusUpdate chan<- FetchStatusUpdate
+	output       chan<- FetchBulkResponse
+}
+
+// New creates new Fetcher. Fetcher needs to be started with Run()
+func New(
+	tempdir string,
+	client *http.Client,
+	input <-chan EndpointFetchRequest,
+	statusUpdate chan<- FetchStatusUpdate,
+	output chan<- FetchBulkResponse,
+) (*Fetcher, error) {
+	f, err := ioutil.TempFile(tempdir, "")
+	if err != nil {
+		return nil, fmt.Errorf("could not create temp zip file in %s: %s", tempdir, err)
+	}
+
+	fetcher := &Fetcher{f, client, input, statusUpdate, output}
+
+	return fetcher, nil
+}
+
+// Run starts fetcher. This method should be run as a gorutine
+func (f *Fetcher) Run(ctx context.Context) {
+	zipWriter := zip.NewWriter(f.file)
+
+	f.workOffRequests(ctx, zipWriter)
+
+	filename := f.file.Name()
+	if err := zipWriter.Close(); err != nil {
+		logrus.WithError(err).Errorf("Could not close zip writer %s", f.file.Name())
```

Comment:
I generally would rather create error messages and bubble them up to the caller, rather than calling them directly deep in a package. If we were to move away from Logrus, for example, we'd have to come into the package to untangle it.

Reply:
Do you think we should put `error` field in response struct?

Reply:
Ya that could work